File: 0001-two-sum/0001-two-sum.py
from typing import List

class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        # A hash map to store the numbers we've seen and their indices.
        # This allows for O(1) average time complexity for lookups.
        num_map = {}

        # We iterate through the list only once.
        # 'enumerate' provides both the index 'i' and the value 'num'.
        for i, num in enumerate(nums):
            # Calculate the 'complement' needed to reach the target.
            complement = target - num

            # Check if the complement already exists in our hash map.
            # If it does, we've found the two numbers that sum up to the target.
            if complement in num_map:
                # The problem guarantees a unique solution, so we can return immediately.
                # 'num_map[complement]' gives the index of the complement.
                return [num_map[complement], i]
            
            # If the complement is not in the map, add the current number and its index.
            # This prepares the map for future lookups.
            num_map[num] = i

        # This part of the code is technically unreachable as the problem statement
        # guarantees that a solution always exists.
        return []

        

File: 0001-two-sum/0001-two-sum.cpp
#include <vector>
#include <unordered_map>

class Solution {
public:
    std::vector<int> twoSum(std::vector<int>& nums, int target) {
        // Use an unordered_map (hash map) for O(1) average time lookups.
        std::unordered_map<int, int> seen_numbers;

        // Iterate through the array once.
        for (int i = 0; i < nums.size(); ++i) {
            int current_num = nums[i];
            int complement = target - current_num;

            // Check if the complement exists in our map.
            if (seen_numbers.count(complement)) {
                // If it does, we found the pair. Return their indices.
                return {seen_numbers[complement], i};
            }
            
            // If not, store the current number and its index for future lookups.
            seen_numbers[current_num] = i;
        }

        // The problem guarantees a solution, so this line is for completeness.
        return {};
    }
};